### PR TITLE
feat: support generic smart contract wallets on open trade

### PR DESCRIPTION
**Pull request by MeLlamoPablo**
## PR Train Visualization

```mermaid
flowchart RL
  a[<a href='https://github.com/prepo-io/prepo-monorepo/pull/1465'>feat/contract-wallet-support</a>] --> main
  d[<a href='https://github.com/prepo-io/prepo-monorepo/pull/1470'>feat/rest-of-tx-approvals</a>] --> a
  b[<a href='https://github.com/prepo-io/prepo-monorepo/pull/1468'>feat/safe-bundle-transactions</a>] --> a
  c[<a href='https://github.com/prepo-io/prepo-monorepo/pull/1469'>chore/deduplicate-safe-sdk</a>] --> b
```

## Summary

This PR adds support for generic smart contract wallets that can't sign permits to the open trade screen. In future PRs, I'll enable support for the rest of the transaction types.

If we detect that the connected wallet is a smart contract, we will fall back to the classic "Approve + Trade" flow

Also in future PRs we will be able to special case Safe{Wallet} and bundle both the approval and the trade in a transaction batch (MultiSend) which is possible if we send the transaction directly [via their SDK](https://safe-global.github.io/safe-core-sdk/#createtransaction) rather than via ethers.

So, to summarize, the ideal state of the app is:

* Regular users have a single-step flow via signing a permit
* Safe{Wallet} users have a single-step flow via batching the transactions
* Generic SC wallet users fall back to the two-step flow

## The introduction of `TransactionBundleStore`

Conceptually, `TransactionBundleStore` can be thought of a store that represents a bundle of transactions. In practice, we will only have bundles of "approval + action" so that is the only case supported here.

`TransactionBundleStore` abstracts away the complexity of dealing with approvals from the UI and the store. It exposes a single `execute` method for the UI to trigger the transaction, and unified `transacting`, `initialLoading` and `actionLabel` properties. Internally, it executes the approval transaction only if necessary, and prompts the user to sign the permit only if necessary.

`TransactionBundleStore` may be overkill as it is, but in future PRs we will expand its functionality to support batching on Safe{Wallet}, so it's good to have this logic contained to a single store.

Note that `TransactionBundleStore` requires unsigned transactions, so this results in some changes:

* `DepositTradeHelper.wrapAndDepositAndTrade` becomes `DepositTradeHelper.createWrapAndDepositAndTradeTx`
* 
* `DepositTradeHelper.tradeForPosition` becomes `DepositTradeHelper.createTradeForPositionTx`
* `Erc20Store.unlockPermanently` becomes `Erc20Store.createUnlockPermanentlyTx`

## How to test

An easy way to test it is deploying a testing Safe{Wallet} and using the [WalletConnect Safe App](https://app.safe.global/share/safe-app?appUrl=https%3A%2F%2Fapps-portal.safe.global%2Fwallet-connect&chain=arb1) to connect to prePO.

### CODE REVIEW

File: apps/frontend/core-dapp/src/features/trade/open-trade/OpenTradeStore.ts [hunk 1]
```diff
@@ -310,74 +320,73 @@ export class OpenTradeStore {
...
+      message = 'Approved Tokens for Trade'
+    } else if (this.paymentToken.type === 'native') {
+      message = 'Deposit & Trade Confirmed'
     } else {
-      toast.success(
-        this.paymentToken.type === 'native' ? 'Deposit & Trade Confirmed' : 'Position Opened',
-        { link: explorerUrl }
-      )
+      message = 'Position Opened'
     }
```
Comment (chrisling-dev) [inline apps/frontend/core-dapp/src/features/trade/open-trade/OpenTradeStore.ts:377]: nit

```
    let message = 'Position Opened'

    if (type === 'approval') {
      message = 'Approved Tokens for Trading'
    } else if (this.paymentToken.type === 'native') {
      message = 'Deposit & Trade Confirmed'
    }
```

**Review by chrisling-dev — CHANGES_REQUESTED** [review]
LGTM after nits, also permit signature is failing with this error, could you look into this please:
Screenshot 2023-07-19 at 1 53 58 PM

**Comment by chrisling-dev** [comment]
@MeLlamoPablo should this be rebased and reopened?

**Comment by MeLlamoPablo** [comment]
> @MeLlamoPablo should this be rebased and reopened?

Yes, working on it

**Comment by MeLlamoPablo** [comment]
@chrisling-dev GitHub won't let me reopen due to the force-push, let's move to #1515

**Pull request closed without merging**